lambda/translmod.ml

```diff
@@ -919,40 +919,134 @@ let required_globals ~flambda body =
+          layout = Pvalue Pgenval;
+          attributes = Lambda.default_param_attribute;
+          mode = Lambda.alloc_heap })
+      params
+  in
+  let inline =
+    (* We want to inline the functor so that [-instantiate] compiles away the
+       function call and actually substitutes. *)
+    Always_inline
+  in
+  lfunction ~kind:(Curried { nlocal = 0 }) ~params ~return:(Pvalue Pgenval)
+    ~attr:{ default_function_attribute with is_a_functor = true; inline }
+    ~loc:Loc_unknown
+    ~body:lam
+    ~mode:alloc_heap
+    ~ret_mode:alloc_heap
+    ~region:true
+
+let transl_implementation_module ~scopes module_id (str, cc, cc2) =
+  let path = global_path module_id in
+  let lam, size =
+    transl_struct ~scopes Loc_unknown [] cc path str
+  in
+  match cc2 with
+  | None -> lam, size, None
+  | Some cc2 -> add_arg_block_to_module_block lam size cc2
+
+let wrap_toplevel_functor_in_struct code =
+  Lprim(Pmakeblock(0, Immutable, None, Lambda.alloc_heap),
+        [ code ],
+        Loc_unknown)
+
+(* Convert an flambda-style implementation (module block only) to a
+   non-flambda-style one (set global as side effect) *)
+
+let wrap_in_setglobal implementation =
+  let code =
+    Lprim (Psetglobal implementation.compilation_unit, [implementation.code],
+           Loc_unknown)
+  in
+  { implementation with code }
+
 (* Compile an implementation *)
 
 type compilation_unit_style =
   | Plain_block
   | Set_global_to_block
   | Set_individual_fields
 
-let transl_implementation_plain_block compilation_unit (str, cc) =
+let has_parameters () =
+  Env.parameters () <> []
+
+let transl_implementation_plain_block compilation_unit impl =
   reset_labels ();
   primitive_declarations := [];
   Translprim.clear_used_primitives ();
   Translcore.clear_probe_handlers ();
   let scopes = enter_compilation_unit ~scopes:empty_scopes compilation_unit in
-  let body, size =
+  let body, (size, arg_block_field) =
     Translobj.transl_label_init (fun () ->
-      let body, size =
-        transl_struct ~scopes Loc_unknown [] cc
-          (global_path compilation_unit) str in
-      Translcore.declare_probe_handlers body, size)
+      let body, size, arg_block_field =
+        transl_implementation_module ~scopes compilation_unit
+          impl
+      in
+      Translcore.declare_probe_handlers body, (size, arg_block_field))
+  in
+  let body, main_module_block_format =
+    match has_parameters () with
+    | false ->
+        body, Mb_record { mb_size = size }
+    | true ->
+        let mb_runtime_params, runtime_param_idents =
+          match Env.runtime_parameters () with
+          | [] ->
+              (* We didn't end up using any of the parameters, but this is still a
+                 parameterised module, so it must still be implemented as a function that
+                 produces a distinct value for each instance. *)
+              let unit_ident = Ident.create_local "*unit*" in
+              [ Rp_unit ], [ unit_ident ]
```

Comment:
Does this correspond to the case where a module is compiled with `-parameter X` but never mentions `X`? I find it strange that this module can be later instantiated without `X`. I think such module should still require `X` for instantiation.

Reply:
As discussed, I'm less worried that this doesn't affect type checking; but I still think having `Unit` in the middle-end complicates things.

Reply:
We discussed later, and because runtime parameters are determined by usage, including real parameters (which I would expected to be determined by registeration), it's difficult to do the way I want. As such, I'm resolving now.

I continue to think that real parameters should be treated seperately from dependencies and determined by registeration.